### PR TITLE
KsqlDB Log Streaming Correction

### PR DESCRIPTION
**Pull request by JumaX**
# Description

This PR fixes a bug whereby KSQL Log Stream functionality is hardcoded and does not support security mechanisms.  Feature can be found here: https://docs.ksqldb.io/en/latest/developer-guide/test-and-debug/processing-log/#log-stream

Specifically, we had to do the following:

1. Add client authentication properties to KSQL log4j Config
2. Have the client connect to the Internal listener
3. Extract the appropriate principals for RBAC depending on the Auth mechanism on the Internal listener (MTLS, KERBEROS, SASL_PLAIN, SASL_SCRAM)
4. Set the appropriate RBAC permissions on the Topic

Log4j does not support OAUTH thus we are unable to use the OAUTH listener for RBAC based clusters, for this feature.  It's mandatory to use the Internal listener with client configurations which match it's Auth Mechanism, as well as the appropriate Topic permissions.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

This bug fix has been validated by running the following Molecule Scenarios Successfully:

1. rbac-mtls-rhel
2. kerberos-rhel
3. rbac-kerberos-debian
4. rbac-scram-custom-rhel

Note: A documentation update is required for RBAC deployments, as it's mandatory for the client principal to exist in LDAP in order for authentication to occur when the KSQL log4j client connects to the internal listener.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 [hunk 1]
```diff
@@ -45,12 +45,67 @@ log4j.appender.kafkaAppender.MaxFileSize={{ksql.kafka_appender_log_file_size}}
 log4j.appender.kafkaAppender.MaxBackupIndex={{ksql.kafka_appender_max_log_files}}
 log4j.appender.kafkaAppender.append=true
 
+{% if ksql_log_streaming_enabled | bool %}
+# Log Streaming Configuration
 log4j.appender.kafka_appender=org.apache.kafka.log4jappender.KafkaLog4jAppender
 log4j.appender.kafka_appender.layout=io.confluent.common.logging.log4j.StructuredJsonLayout
-log4j.appender.kafka_appender.BrokerList=localhost:9092
-log4j.appender.kafka_appender.Topic=default_ksql_processing_log
+log4j.appender.kafka_appender.BrokerList={% for host in groups['kafka_broker'] %}{% if loop.index > 1%},{% endif %}{{ host }}:{{kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['port']}}{% endfor %}
```
Comment (domenicbove) [inline roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:52]: why wouldnt this be the same listener ksql goes to kafka?

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 [hunk 2]
```diff
@@ -45,12 +45,67 @@ log4j.appender.kafkaAppender.MaxFileSize={{ksql.kafka_appender_log_file_size}}
...
+log4j.appender.kafka_appender.Topic={{ksql_processing_log}}
 log4j.logger.processing=ERROR, kafka_appender
 
+# Security Configuration for Log Streaming
+{% set config_prefix = '' %}
```
Comment (domenicbove) [inline roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:57]: it appears the config_prefix var does not need to exist

File: roles/confluent.variables_handlers/defaults/main.yml [hunk 3]
```diff
@@ -262,12 +262,16 @@ ksql:
 ksql_jolokia_enabled: "{{jolokia_enabled}}"
 ksql_jolokia_port: 7774
 ksql_jolokia_ssl_enabled: "{{ ksql_ssl_enabled }}"
+ksql_processing_log: "processing_log"
+ksql_log_streaming_enabled: false
```
Comment (domenicbove) [inline roles/confluent.variables_handlers/defaults/main.yml:266]: should this be turned on by default? i think thats the ksql recommendation

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 (hunk 1 above)
Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:52]: Good question.  To keep it short, if you are using RBAC, it won't work.  log4j does not support OAuth at this time, so you have to use another listener to connect on.  There is a documentation update I shared over slack that I asked for the team's review on, which steps through this.

File: roles/confluent.variables_handlers/defaults/main.yml (hunk 3 above)
Comment (JumaX) [reply roles/confluent.variables_handlers/defaults/main.yml:266]: I spoke with Product about this, and due to the way that Log Streaming works with RBAC, Product made the decision to disable it by default as it adds far more overhead to the end user to setup.  The idea being, giving the user the choice to go through that overhead if they really want this feature.  It's directly related to my previous comment about the listener and the document I shared on slack.

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 (hunk 2 above)
Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2]: Good catch!  It's been removed.

File: roles/confluent.ksql/defaults/main.yml [hunk 4]
```diff
@@ -50,3 +50,5 @@ ksql:
     ksql.streams.producer.max.block.ms: 9223372036854775807
     producer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor
     consumer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor
+    ksql.logging.processing.topic.auto.create: true
```
Comment (domenicbove) [inline roles/confluent.ksql/defaults/main.yml:53]: do you want these properties if ksql_log_streaming_enabled == false?

File: roles/confluent.ksql/tasks/rbac.yml [hunk 5]
```diff
@@ -1,4 +1,8 @@
 ---
+- name: Get Principal used on the Inter Broker Listener
+  include_tasks: set_principal.yml
+  when: ksql_log_streaming_enabled | bool
+
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/rbac.yml:5]: The set_principal.yml file looks like duplicating code, itd be cleaner to do this:

```suggestion
- name: Get Principal used on the Inter Broker Listener
  include_role: 
    name: confluent.kafka_broker
    tasks_from: set_principal.yml
  when: ksql_log_streaming_enabled | bool

- set_fact:
    ksql_log4j_principal: kafka_broker_principal
  when: ksql_log_streaming_enabled | bool
```

Comment (domenicbove) [reply roles/confluent.ksql/tasks/rbac.yml:5]: or there isnt a need for the ksql_log4j_principal variable, you could just use the var kafka_broker_principal down below

File: roles/confluent.variables_handlers/defaults/main.yml (hunk 3 above)
Comment (domenicbove) [reply roles/confluent.variables_handlers/defaults/main.yml:266]: Ah yes that makes sense about rbac and the listener because the interbroker listener can't be oauth. 

Seems to me a simpler solution would be just to not have ksql_log_streaming_enabled at all if rbac is enabled:
```suggestion
ksql_log_streaming_enabled: "{{ not rbac_enabled }}"
```

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 [hunk 6]
```diff
@@ -45,12 +45,66 @@ log4j.appender.kafkaAppender.MaxFileSize={{ksql.kafka_appender_log_file_size}}
...
+{% set keystore_storepass = ksql_keystore_storepass %}
+{% set keystore_keypass = ksql_keystore_keypass %}
+{% set sasl_plain_username = sasl_plain_users.ksql.principal %}
+{% set sasl_plain_password = sasl_plain_users.ksql.password %}
+{% set sasl_scram_username = sasl_scram_users.ksql.principal %}
+{% set sasl_scram_password = sasl_scram_users.ksql.password %}
+{% set kerberos_keytab_path = ksql_kerberos_keytab_path %}
+{% set kerberos_principal = ksql_kerberos_principal %}
+{% set oauth_username = ksql_ldap_user %}
+{% set oauth_password = ksql_ldap_password %}
```
Comment (domenicbove) [inline roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:70]: are these set lines necessary? you could use what the variables are set to below.

i would keep the 
`{% set listener = kafka_broker_listeners[kafka_broker_inter_broker_listener_name] %}`
line, it improves readablility below

File: roles/confluent.ksql/tasks/rbac.yml (hunk 5 above)
Comment (domenicbove) [reply roles/confluent.ksql/tasks/rbac.yml:5]: I just realized to make this work in the referenced set_principal.yml it would need the ability to pass a keystore path variable, so like this:
```
- name: Extract Common Name from Keystore - SSL Mutual Auth
  # Examine the keystore
  # Search lines with Entry type: "PrivateKeyEntry" and return that line and all after, ca cert is of type "trustedCertEntry"
  # Search for first "Owner" line
  # Extract DNAME from line
  # Remove spaces after commas
  shell: |
    keytool -list -keystore {{keystore_path | default(kafka_broker_keystore_path)}} \
```

and then in your include_role, it would be:
```
- name: Get Principal used on the Inter Broker Listener
  include_role: 
    name: confluent.kafka_broker
    tasks_from: set_principal.yml
  vars:
    keystore_path: ksql_keystore_path
  when: ksql_log_streaming_enabled | bool
```

File: roles/confluent.variables_handlers/defaults/main.yml (hunk 3 above)
Comment (JumaX) [reply roles/confluent.variables_handlers/defaults/main.yml:266]: @domenicbove I've spoken with the KSQL team and this feature is expected to function in this way with RBAC.  As in, this feature should be available even when RBAC is enabled and use another listener.

File: roles/confluent.ksql/defaults/main.yml (hunk 4 above)
Comment (JumaX) [reply roles/confluent.ksql/defaults/main.yml:53]: Good catch!

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 (hunk 6 above)
Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:70]: Fair.

File: roles/confluent.ksql/tasks/rbac.yml (hunk 5 above)
Comment (JumaX) [reply roles/confluent.ksql/tasks/rbac.yml:5]: @domenicbove Happy to discuss this further on a call, but initially, I disagree.  I actually did think about this, however came to the conclusion that this should be self contained to the KSQL role, as it is such a specific piece of functionality to KSQL setup due to the unique configuration of this feature.  I think sharing this code could lead to serious problems if it's edited in the future and the context of where it's used is not fully understood.

It generally falls in line with Ansible best practices as well, whereby its recommended roles are self contained.  I say this with the understanding that we already don't follow this best practice in a lot of places. :). I think however, due to how quirky log streaming is, we should keep it separate.  Happy to discuss further though. :)

File: roles/confluent.variables_handlers/defaults/main.yml (hunk 3 above)
Comment (domenicbove) [reply roles/confluent.variables_handlers/defaults/main.yml:266]: Dang, really wish we could just say "no ksql log streaming if you have rbac on", because it just feels so unnatural to connect to the interbroker listener instead of the ksql listener. 

Wasn't there a whole security patch that said if rbac is on each component must connect over oauth?
I remember the operator team had to struggle w that and we were better off bc we didn't allow that.

File: roles/confluent.ksql/defaults/main.yml [hunk 7]
```diff
@@ -50,3 +50,5 @@ ksql:
     ksql.streams.producer.max.block.ms: 9223372036854775807
     producer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor
     consumer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor
+    ksql.logging.processing.topic.auto.create: "{{ksql_log_streaming_enabled}}"
+    ksql.logging.processing.topic.name: "{{ksql_service_id}}{{ksql_processing_log}}"
```
Comment (domenicbove) [inline roles/confluent.ksql/defaults/main.yml:54]: Should these be moved into the properties template with an if statement around them?
Also I think you'll need the `lower` filter:
```
ksql.logging.processing.topic.auto.create={{ksql_log_streaming_enabled|lower}}
```

File: roles/confluent.ksql/tasks/set_principal.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
+---
+- name: Set Principal - Sasl Scram
+  set_fact:
+    ksql_log4j_principal: "{{ sasl_scram_users.admin.principal }}"
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/set_principal.yml:4]: Should ksql be acting as the kafka admin or the scram user set up for ksql?
sasl_scram_users.admin.principal or sasl_scram_users.ksql.principal

File: roles/confluent.variables_handlers/defaults/main.yml (hunk 3 above)
Comment (JumaX) [reply roles/confluent.variables_handlers/defaults/main.yml:266]: Completely agree with you!  It maybe worth discussing this again with Product to get a final say on how to proceed here.  A few other engineers have also been surprised with the configuration around this as well.

File: roles/confluent.ksql/defaults/main.yml (hunk 7 above)
Comment (JumaX) [reply roles/confluent.ksql/defaults/main.yml:54]: We could but it doesn't really impact anything as nothing is done unless the feature is enabled.

Comment (JumaX) [reply roles/confluent.ksql/defaults/main.yml:54]: I'll move it.

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 [hunk 9]
```diff
@@ -44,13 +44,57 @@ log4j.appender.kafkaAppender.layout.ConversionPattern=[%d] %p %m (%c:%L)%n
 log4j.appender.kafkaAppender.MaxFileSize={{ksql.kafka_appender_log_file_size}}
 log4j.appender.kafkaAppender.MaxBackupIndex={{ksql.kafka_appender_max_log_files}}
 log4j.appender.kafkaAppender.append=true
+{% if ksql_log_streaming_enabled | bool %}
+ksql.logging.processing.topic.auto.create: {{ksql_log_streaming_enabled | lower}}
```
Comment (domenicbove) [inline roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:48]: These need to get into ksql-server.properties (or whatever the filename is) not the log4j props right?

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 [hunk 10]
```diff
@@ -44,13 +44,57 @@ log4j.appender.kafkaAppender.layout.ConversionPattern=[%d] %p %m (%c:%L)%n
...
 log4j.appender.kafka_appender.layout=io.confluent.common.logging.log4j.StructuredJsonLayout
-log4j.appender.kafka_appender.BrokerList=localhost:9092
-log4j.appender.kafka_appender.Topic=default_ksql_processing_log
+log4j.appender.kafka_appender.BrokerList={% for host in groups['kafka_broker'] %}{% if loop.index > 1%},{% endif %}{{ host }}:{{kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['port']}}{% endfor %}
+
+log4j.appender.kafka_appender.Topic={{ksql_processing_log}}
 log4j.logger.processing=ERROR, kafka_appender
 
+# Security Configuration for Log Streaming
+{% set listener = kafka_broker_listeners[kafka_broker_inter_broker_listener_name] %}
```
Comment (domenicbove) [inline roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:60]: maybe the listener should only be the interbroker listener in an rbac enabled situation. if rbac is disabled it would make more sense to use ksql listener, wdyt?

I would create a var like:
```
ksql_processing_log_listener: "{{kafka_broker_inter_broker_listener_name if rbac_enabled else ksql_listener}}"
```

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 (hunk 9 above)
Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:48]: You're right... this is what I get for editing late at night.

File: roles/confluent.ksql/templates/ksql-server_log4j.properties.j2 (hunk 10 above)
Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:60]: We enable two listeners by default in every setup we do, and we recommend not to modify this, unless adding additional listeners.  So provided the user sticks to our supported configuration, there shouldn't be a need for this change.

Comment (domenicbove) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:59]: So up until this PR we have components that talk to kafka over their designated listener, and there is a var naming which listener to use: https://github.com/confluentinc/cp-ansible/blob/5.5.x/roles/confluent.variables_handlers/defaults/main.yml#L257

One would expect that all parts of ksql would connect to kafka over the same listener, but alas that does not work in the case of the processing log WHEN rbac is enabled. My point is in a non rbac installation I would expect the interbroker listener to only be used for interbroker connections.

My suggestion is to create a new variable under line 257 like so:
```
ksql_processing_log_kafka_listener_name: "{{kafka_broker_inter_broker_listener_name if rbac_enabled else ksql_kafka_listener_name}}"
```

And then use that variable everywhere in this pr instead of `kafka_broker_inter_broker_listener_name`

Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:59]: Really good point.  I am wondering though, is it more or less confusing for this feature to always connect in the same way (current implementation), or to switch depending on the auth mechanism (your suggestion).  What makes more sense to the end user?

Comment (JumaX) [reply roles/confluent.ksql/templates/ksql-server_log4j.properties.j2:59]: To confirm our conversation, using a variable would the user the opportunity to override the listener used for this feature if they do not want to use the inter-broker listener for security reasons.

If we replace inter-broker listener with ksql_processing_log_kafka_listener_name: variable should give the flexibility.

File: roles/confluent.ksql/tasks/rbac.yml [hunk 11]
```diff
@@ -85,6 +89,11 @@
             "resourceType":"Group",
             "name":"_confluent-ksql-{{ksql.properties['ksql.service.id']}}",
             "patternType":"PREFIXED"
+          },
+          {
+            "resourceType":"Topic",
+            "name":"{{ksql.properties['ksql.service.id']}}processing_log",
+            "patternType":"LITERAL"
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/rbac.yml:96]: is this block necessary? looks like the rolebinding is conditionally created below?

File: roles/confluent.ksql/tasks/main.yml [hunk 12]
```diff
@@ -60,7 +60,7 @@
     kerberos_user: "{{ksql_user}}"
     kerberos_keytab_path: "{{ksql_kerberos_keytab_path}}"
     kerberos_handler: "restart ksql"
-  when: kafka_broker_listeners[ksql_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI'
+  when: (kafka_broker_listeners[ksql_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI') or (kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI' and ksql_log_streaming_enabled | bool)
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/main.yml:63]: kafka_broker_inter_broker_listener_name -> ksql_processing_log_kafka_listener_name

File: roles/confluent.ksql/tasks/main.yml [hunk 13]
```diff
@@ -60,7 +60,7 @@
     kerberos_user: "{{ksql_user}}"
     kerberos_keytab_path: "{{ksql_kerberos_keytab_path}}"
     kerberos_handler: "restart ksql"
-  when: (kafka_broker_listeners[ksql_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI') or (kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI' and ksql_log_streaming_enabled | bool)
+  when: (kafka_broker_listeners[ksql_processing_log_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI') or (kafka_broker_listeners[ksql_processing_log_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI' and ksql_log_streaming_enabled | bool)
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/main.yml:63]:
```suggestion
  when: (kafka_broker_listeners[ksql_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI') or (kafka_broker_listeners[ksql_processing_log_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI' and ksql_log_streaming_enabled | bool)
```

File: roles/confluent.test/molecule/rbac-kerberos-debian/molecule.yml [hunk 14]
```diff
@@ -209,6 +209,10 @@ provisioner:
           ldap.group.member.attribute.pattern=CN=(.*),OU=rbac,DC=example,DC=com
           ldap.user.object.class=account
 
+      ksql:
```
Comment (domenicbove) [inline roles/confluent.test/molecule/rbac-kerberos-debian/molecule.yml:212]: does this test validate ksql processing logs? `ksql_log_streaming_enabled` defaults to false, i think you need to turn it on and add verify tasks

Comment (JumaX) [reply roles/confluent.test/molecule/rbac-kerberos-debian/molecule.yml:212]: When we had discussed this previously, you stated that you were fine with me implementing this and validating that it works.  I can write additional test, but that wasn't what was previously discussed thus why its not here.

Comment (domenicbove) [reply roles/confluent.test/molecule/rbac-kerberos-debian/molecule.yml:212]: If you want to go the route of not adding any tests for this feature then i dont think there needs to be any changes to this file

but bc there is such complex rbac/kerberos code it'd probably be best to add a test

File: roles/confluent.ksql/tasks/main.yml [hunk 15]
```diff
@@ -60,7 +60,7 @@
     kerberos_user: "{{ksql_user}}"
     kerberos_keytab_path: "{{ksql_kerberos_keytab_path}}"
     kerberos_handler: "restart ksql"
-  when: kafka_broker_listeners[ksql_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI'
+  when: (kafka_broker_listeners[ksql_processing_log_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI') or (kafka_broker_listeners[ksql_processing_log_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI' and ksql_log_streaming_enabled | bool)
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/main.yml:63]:
```suggestion
  when: (kafka_broker_listeners[ksql_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI') or (kafka_broker_listeners[ksql_processing_log_kafka_listener_name]['sasl_protocol'] | default(sasl_protocol) | normalize_sasl_protocol == 'GSSAPI' and ksql_log_streaming_enabled | bool)
```

File: roles/confluent.ksql/tasks/rbac.yml [hunk 16]
```diff
@@ -144,3 +148,32 @@
       }
     status_code: 204
   when: "'schema_registry' in groups"
+
+- name: Grant ksql user the ResourceOwner role on the Processing Log Topic
+  uri:
+    url: "{{mds_http_protocol}}://{{ groups['kafka_broker'][0] }}:{{mds_port}}/security/1.0/principals/User:{{ksql_ldap_user}}/roles/ResourceOwner/bindings"
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/rbac.yml:154]:
```suggestion
    url: "{{mds_http_protocol}}://{{ groups['kafka_broker'][0] }}:{{mds_port}}/security/1.0/principals/User:{{ksql_log4j_principal}}/roles/ResourceOwner/bindings"
```

is this supposed to be the log4j principal or the ksql ldap user?

Comment (JumaX) [reply roles/confluent.ksql/tasks/rbac.yml:154]: Yep, but I keep having to re-edit things due to previous changes, and my changes keeping getting munged.  If you notice it was correct previously.  Honestly, I'm not sure why this keeps happening.  I fix something and then something else gets committed.  I've been using stash's and I think that's where the confusion is coming from.

Comment (JumaX) [reply roles/confluent.ksql/tasks/rbac.yml:183]: So I figured out what is happening.  You need to grant Resource Owner on the log streaming topic to both users.

The ksql_ldap_user, is the user KSQL runs as via it's OAuth connection, and is what is used to create the topic when ksql is started and log streaming is enabled.

The ksql_log4j_principal, is the user we extract from the keytab in this instance, or TLS cert etc, that is used to create the client connection and write to the log streaming topic, thus it also needs to be a resource owner.

So both users need these permissions when log streaming is enabled.

I've tested this by only having the ksql_log4j_principal set, and KSQL fails to start, because it is unable to create the Topic, it actually throws an error stating it can't authorize to create it.

TL:DR. You need both users to have access to the log streaming topic, when log streaming is enabled.

Comment (domenicbove) [reply roles/confluent.ksql/tasks/rbac.yml:183]: copy that! nice investigation

File: roles/confluent.ksql/tasks/rbac.yml [hunk 17]
```diff
@@ -144,3 +148,61 @@
...
+        "resourcePatterns": [
+          {
+            "resourceType":"Topic",
+            "name":"{{ksql.properties['ksql.service.id']}}processing_log",
+            "patternType":"LITERAL"
+          }
+        ]
+      }
+    status_code: 204
+  when: ksql_log_streaming_enabled | bool
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/rbac.yml:208]:
```suggestion
- name: Grant ksql user the ResourceOwner role on the Processing Log Topic
  uri:
    url: "{{mds_http_protocol}}://{{ groups['kafka_broker'][0] }}:{{mds_port}}/security/1.0/principals/User:{{item}}/roles/ResourceOwner/bindings"
    method: POST
    validate_certs: false
    force_basic_auth: true
    url_username: "{{mds_super_user}}"
    url_password: "{{mds_super_user_password}}"
    headers:
      Content-Type: application/json
    body_format: json
    body: >
      {
        "scope": {
          "clusters": {
            "kafka-cluster":"{{kafka_cluster_id}}"
          }
        },
        "resourcePatterns": [
          {
            "resourceType":"Topic",
            "name":"{{ksql.properties['ksql.service.id']}}processing_log",
            "patternType":"LITERAL"
          }
        ]
      }
    status_code: 204
    loop:
      - "{{ksql_ldap_user}}"
      - "{{ksql_log4j_principal}}"
    when: ksql_log_streaming_enabled | bool
```

Should be able to cut down some lines by using loop

File: roles/confluent.ksql/tasks/rbac.yml [hunk 18]
```diff
@@ -144,3 +148,35 @@
...
+            "name":"{{ksql.properties['ksql.service.id']}}processing_log",
+            "patternType":"LITERAL"
+          }
+        ]
+      }
+    status_code: 204
+    loop:
+      - "{{ksql_ldap_user}}"
+      - "{{ksql_log4j_principal}}"
+    when: ksql_log_streaming_enabled | bool
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/rbac.yml:182]:
```suggestion
  loop:
    - "{{ksql_ldap_user}}"
    - "{{ksql_log4j_principal}}"
  when: ksql_log_streaming_enabled | bool
```
looks like the indentation needs to go one back

**Review by domenicbove — APPROVED** [review]
nice work!
